kernel/aster-nix/src/fs/utils/page_cache.rs

```diff
@@ -85,16 +87,179 @@ impl Debug for PageCache {
     }
 }
 
+struct ReadaheadWindow {
+    /// Window start.
+    start: usize,
+    /// Window size.
+    size: usize,
+    /// Look ahead position in current window.
+    lookahead_index: usize,
+}
+
+impl ReadaheadWindow {
+    pub fn new(start_page: usize, init_size: usize, max_page: usize) -> Option<Self> {
+        if start_page >= max_page {
+            return None;
+        }
+        Some(Self {
+            start: start_page,
+            size: init_size.min(max_page - start_page),
+            lookahead_index: start_page,
+        })
+    }
+
+    pub fn update(&self, max_size: usize, max_page: usize) -> Option<Self> {
+        let new_start = self.start + self.size;
+        if new_start >= max_page {
+            return None;
+        }
+        let new_size = (self.size * 2).min(max_size).min(max_page - new_start);
+        Some(Self {
+            start: new_start,
+            size: new_size,
+            lookahead_index: new_start,
+        })
+    }
+
+    pub fn lookahead_index(&self) -> usize {
+        self.lookahead_index
+    }
+
+    pub fn readahead_index(&self) -> usize {
+        self.start + self.size
+    }
+
+    pub fn ra_range(&self) -> Range<usize> {
+        self.start..self.start + self.size
+    }
+}
+
+struct ReadaheadState {
+    /// Current readahead window.
+    ra_window: Option<ReadaheadWindow>,
+    /// Maximum window size.
+    max_size: usize,
+    /// Last page visited.
+    prev_page: Option<usize>,
+    /// Readahead requests waiter.
+    waiter: BioWaiter,
+}
+
+impl ReadaheadState {
+    const INIT_WINDOW_SIZE: usize = 4;
+    const MAX_WINDOW_SIZE: usize = 32;
+
+    pub fn new() -> Self {
+        Self {
+            ra_window: None,
+            max_size: Self::MAX_WINDOW_SIZE,
+            prev_page: None,
+            waiter: BioWaiter::new(),
+        }
+    }
+
+    /// Set the maximum readahead window size.
+    pub fn set_max_window_size(&mut self, size: usize) {
+        self.max_size = size.min(Self::MAX_WINDOW_SIZE).max(Self::INIT_WINDOW_SIZE);
+    }
+
+    fn is_sequential(&self, idx: usize) -> bool {
+        if let Some(prev) = self.prev_page {
+            idx == prev || idx == prev + 1
+        } else {
+            false
+        }
+    }
+
+    fn has_completed_all_requests(&self) -> bool {
+        for i in 0..self.waiter.nreqs() {
+            if self.waiter.status(i) == BioStatus::Submit {
+                return false;
+            }
+        }
+        true
+    }
+
+    pub fn wait_for_prev_readahead(
+        &mut self,
+        pages: &mut MutexGuard<LruCache<usize, Page>>,
+        force: bool,
```

Comment:
The force parameter is used to indicate whether to wait for the current request to complete. Maybe because I have written some logic that does not belong to this function, making this function a little confused. I've removed that part of the logic.

Reply:
I suggest to rewite is as this:
```rust
pub fn wait_for_prev_readahead(
    &mut self,
    pages: &mut MutexGuard<LruCache<usize, Page>>,
) -> Result<()> {
    if matches!(self.waiter.wait(), Some(BioStatus::Complete)) {
        let Some(window) = &self.ra_window else {
            return_errno!(Errno::EINVAL);
        };
        for idx in window.ra_range() {
            if let Some(page) = pages.get_mut(&idx) {
                page.set_state(PageState::UpToDate);
            }
        }
        self.waiter.clear();
    } else {
        return_errno!(Errno::EIO);
    }

    Ok(())
}


fn ondemand_readahead(&self, idx: usize) -> Result<Frame> {
    let mut pages = self.pages.lock();
    let mut ra_state = self.ra_state.lock();
    let backend = self.backend();
    // Check for the previous readahead.
    if ra_state.has_completed_all_requests() {
        ra_state.wait_for_prev_readahead(&mut pages)?;
    }

    ...
    ...
    
    if let PageState::Uninit = page.state() {
        // Cond 2: We should wait for the previous readahead.
        // If there is no previous readahead, an error must have occurred somewhere.
        if ra_state.request_number() == 0 {
            return_errno!(Errno::EINVAL)
        }
        ra_state.wait_for_prev_readahead(&mut pages)?;
        pages.get(&idx).unwrap().frame().clone()


```